pilot/pkg/status/state.go

```diff
@@ -0,0 +1,307 @@
+	// the update() function is not implemented, and the Get() function returns the resource
+	// in a different format than is needed for k8s.updateStatus.
+	c.currentlyWriting.Lock(config)
+	defer c.currentlyWriting.Unlock(config)
+	resourceInterface := c.initK8sResource(config.GroupVersionResource).
+		Namespace(config.Namespace)
+	// should this be moved to some sort of InformerCache for speed?
+	current, err := resourceInterface.Get(ctx, config.Name, metav1.GetOptions{ResourceVersion: config.ResourceVersion})
+	if err != nil {
+		if errors.IsGone(err) || errors.IsNotFound(err) {
+			// this resource has been deleted.  prune its state and move on.
+			c.pruneOldVersion(config)
+			return
+		}
+		scope.Errorf("Encountered unexpected error when retrieving status for %v: %s", config, err)
+		return
+
+	}
+	if config.ResourceVersion != current.GetResourceVersion() {
+		// this distribution report is for an old version of the object.  Prune and continue.
+		c.pruneOldVersion(config)
+		return
+	}
+	// check if status needs updating
+	if needsReconcile, desiredStatus := ReconcileStatuses(current.Object, distributionState, c.clock); needsReconcile {
+		// technically, we should be updating probe time even when reconciling isn't needed, but
+		// I'm skipping that for efficiency.
+		current.Object["status"] = desiredStatus
+		_, err := resourceInterface.UpdateStatus(ctx, current, metav1.UpdateOptions{})
+		if err != nil {
+			scope.Errorf("Encountered unexpected error updating status for %v, will try again later: %s", config, err)
+			return
+		}
+	}
+}
+
+func (c *DistributionController) pruneOldVersion(config Resource) {
+	defer c.mu.Unlock()
+	c.mu.Lock()
+	delete(c.CurrentState, config)
+}
+
+func (c *DistributionController) removeStaleReporters(staleReporters []string) {
+	defer c.mu.Unlock()
+	c.mu.Lock()
+	for key, fractions := range c.CurrentState {
+		for _, staleReporter := range staleReporters {
+			delete(fractions, staleReporter)
+		}
+		c.CurrentState[key] = fractions
+	}
+}
+
+func getTypedStatus(in interface{}) (out IstioStatus, err error) {
+	var statusBytes []byte
+	if statusBytes, err = yaml.Marshal(in); err == nil {
+		err = yaml.Unmarshal(statusBytes, &out)
+	}
+	return
+}
+
+func boolToConditionStatus(b bool) v1beta1.ConditionStatus {
+	if b {
+		return metav1.ConditionTrue
+	}
+	return metav1.ConditionFalse
+}
+
+func ReconcileStatuses(current map[string]interface{}, desired Progress, clock clock.Clock) (bool, *IstioStatus) {
+	needsReconcile := false
+	currentStatus, err := getTypedStatus(current["status"])
+	desiredCondition := IstioCondition{
+		Type:               Reconciled,
+		Status:             boolToConditionStatus(desired.AckedInstances == desired.TotalInstances),
+		LastProbeTime:      metav1.NewTime(clock.Now()),
+		LastTransitionTime: metav1.NewTime(clock.Now()),
+		Message:            fmt.Sprintf("%d/%d proxies up to date.", desired.AckedInstances, desired.TotalInstances),
```

Comment:
This is still 4x the correct count?

Reply:
Looking at this more, I thought we were continuing to update this as new proxies came up. But it seems like we actually don't, so the message is not very accurate. It may say 12/12 proxies updated when I may have 1000s of proxies in the cluster (assuming you create config, then spin up new proxies, in this example).

I actually think that is what it **should** do, but if we are going to do that I feel like it should not claim any numbers here - that can be internal only.

Reply:
Regarding the 4x, this was the original design (and is how istioctl wait works even now), but I think we can do better.  Maybe we count each envoy only once, and consider it as "not yet acked" until all of it's subscriptions have been updated?

Regarding updating as new proxies come online - I think you are correct.  It's not particularly relevant to the config when a new envoy has sync'ed, particularly because that envoy would have never received the 'old' config, so even if your change is security critical, a 'not yet acked' envoy proxy doesn't represent a concern.  The purpose of exposing the count was to allow users in very large clusters to proceed when a threshold of envoys have acked.  Perhaps we should represent this as a percent instead?

Reply:
Any number is inaccurate though, because it stops updating. If they want a percent only, then perhaps it should be a setting. We expose binary reconcile/not reconciled, and they can set the threshold where that happens. or maybe for the initial change we consider that out of scope and see if users complain about the binary. I just feel like reporting something that isn't accurate like "5/5 proxies up to date" is not a great idea.

Reply:
We considered allowing a cluster wide threshold setting, but avoided that because some changes will likely have different thresholds than others.  If we move to a percent system, and clarify in the documents that 100% means that none of the proxies have a config that is older than that, I think that could meet the truth in reporting standard.  What do you think?  Should we have an extended talk about this next week, and push any changes as a bugfix, or should we figure it out here and now?